### PR TITLE
[tune] Add `points_to_evaluate` to BasicVariantGenerator

### LINKED ISSUE #9159: [tune] Ray-Tune Explicit Hyperparameter Initialization
### What is your question?

Hello, I have a question about Ray-Tune hyperparameter initialization. 

This is a piece of code copied from the official [Ray-Tune tutorial](https://docs.ray.io/en/master/tune/tutorials/tune-advanced-tutorial.html).
```
    analysis = tune.run(
        PytorchTrainble,
        name="pbt_test",
        scheduler=scheduler,
        reuse_actors=True,
        verbose=1,
        stop=stopper,
        export_formats=[ExportFormat.MODEL],
        checkpoint_score_attr="mean_accuracy",
        checkpoint_freq=5,
        keep_checkpoints_num=4,
        num_samples=4,
        config={
            "lr": tune.uniform(0.001, 1),
            "momentum": tune.uniform(0.001, 1),
        })
```

We specified the population size to be 4, each of the population would sample initial hyperparameters randomly for lr and momemtum from uniform distribution (0.001, 1). I wonder if there is any way to explicitly specify the initialization. For example, I would like the initial hyperparameters for the 4 samples to be:

[{"lr": 0.1, "momentum": 0.9}, {"lr": 0.2, "momentum": 0.8}, {"lr": 0.3, "momentum": 0.7}, {"lr": 0.4, "momentum": 0.6},]

Thank you.

*Ray version and other system information (Python version, TensorFlow version, OS):*
Ray 0.9.0dev
Python 3.6

### PR DESCRIPTION
**Pull request by krfricke**
## Why are these changes needed?

We support `points_to_evaluate` for all search algorithms except for the basic variant generator (random and grid search). This PR adds support for this.

Notably, the `points_to_evaluate` passed to the `BasicVariantGenerator` can be partially specified. E.g. if we have a config like this:

```
config = {
    "a": tune.uniform(1.0, 4.0),
    "b": tune.grid_search(["Alpha", "Beta"]),
    "c": tune.sample_from(lambda spec: spec.config["a"] * -1),
}
```
we can specify initial points like this:
```
search_alg = tune.suggest.BasicVariantGenerator(points_to_evaluate=[
    {
        "a": 3.0,
        "b": "Alpha"
    },
    {
        "c": -7.0
    }
])
```

And the unspecified values will be sampled as before. This is also true for grid search variables. Generally, each point passed in `points_to_evaluate` will replace exactly one `num_samples` sample. If a grid search variable is overwritten like this, the number of total samples thus reduces to one for this sample (instead of the number of grid search values).

## Related issue number

Closes https://github.com/ray-project/ray/issues/9159

## Checks

- [x] I've run `scripts/format.sh` to lint the changes in this PR.
- [x] I've included any doc changes needed for https://docs.ray.io/en/master/.
- [x] I've made sure the tests are passing. Note that there might be a few flaky tests, see the recent failures at https://flakey-tests.ray.io/
- Testing Strategy
   - [x] Unit tests
   - [ ] Release tests
   - [ ] This PR is not tested :(

### CODE REVIEW

File: python/ray/tune/suggest/variant_generator.py [hunk 1]
```diff
@@ -172,6 +198,39 @@ def _generate_variants(spec: Dict) -> Tuple[Dict, Dict]:
             yield resolved_vars, spec
 
 
+def get_preset_variants(spec: Dict, config: Dict):
+    """Get variants according to a spec, but first overwrite spec config
+    variables with values from `config`"""
```
Comment (richardliaw) [inline python/ray/tune/suggest/variant_generator.py:203]: btw, generally, the first line of the docstring should fit on one line (no line wrap)

File: python/ray/tune/suggest/basic_variant.py [hunk 2]
```diff
@@ -16,6 +18,25 @@ class BasicVariantGenerator(SearchAlgorithm):
...
+    ``num_samples=5``, and if ``a=2`` is passed as the first and only
+    ``points_to_evaluate``, a total number of 13 samples is generated: One
+    for ``a=2``, and then 4 samples for each of the 3 grid search variables.
+
+    Args:
+        points_to_evaluate (list): Initial parameter suggestions to be run
+        first. This is for when you already have some good parameters
+        you want to run first to help the algorithm make better suggestions
+        for future parameters. Needs to be a list of dicts containing the
+        configurations.
```
Comment (richardliaw) [inline python/ray/tune/suggest/basic_variant.py:38]:
```suggestion
        points_to_evaluate (list): Initial parameter suggestions to be run
            first. This is for when you already have some good parameters
            you want to run first to help the algorithm make better suggestions
            for future parameters. Needs to be a list of dicts containing the
            configurations.
```

**Review by richardliaw — APPROVED** [review]
Looks good! 

I am not actually sure if this shows up in the docs, but perhaps it might be good to make sure this is referenceable from the user guide.

File: python/ray/tune/suggest/variant_generator.py (hunk 1 above)
Comment (krfricke) [reply python/ray/tune/suggest/variant_generator.py:203]: In this case I ignored this as we're not documenting this internal function anywhere. It was just for internal reference. I'll still fix it

**Review by amogkam — APPROVED** [review]
LGTM!

**Comment by krfricke** [comment]
I just added a bit more comprehensive documentation.

File: python/ray/tune/examples/tune_basic_example.py [hunk 3]
```diff
@@ -0,0 +1,51 @@
...
+    # The `width` and `height` parameters are sampled randomly.
+    # `steps` is a constant parameter.
+
+    analysis = tune.run(
+        easy_objective,
+        metric="mean_loss",
+        mode="min",
+        num_samples=5 if args.smoke_test else 50,
+        config={
+            "steps": 100,
```
Comment (richardliaw) [inline python/ray/tune/examples/tune_basic_example.py:45]:
```suggestion
            "steps": 5 if args.smoke_test else 100,
```

**Review by richardliaw — APPROVED** [review]

**Pull request merged**